In [1]:
import numpy as np
import pandas as pd


In [2]:
df = pd.read_csv('dataset.csv')

In [3]:
df.head()

Release Year                             Title Origin/Ethnicity  \
0          1901            Kansas Saloon Smashers         American   
1          1901     Love by the Light of the Moon         American   
2          1901           The Martyred Presidents         American   
3          1901  Terrible Teddy, the Grizzly King         American   
4          1902            Jack and the Beanstalk         American   

                             Director Cast    Genre  \
0                             Unknown  NaN  unknown   
1                             Unknown  NaN  unknown   
2                             Unknown  NaN  unknown   
3                             Unknown  NaN  unknown   
4  George S. Fleming, Edwin S. Porter  NaN  unknown   

                                           Wiki Page  \
0  https://en.wikipedia.org/wiki/Kansas_Saloon_Sm...   
1  https://en.wikipedia.org/wiki/Love_by_the_Ligh...   
2  https://en.wikipedia.org/wiki/The_Martyred_Pre...   
3  https://en.wikipedia.org/wiki/Terrible_Teddy,_...   
4  https://en.wikipedia.org/wiki/Jack_and_the_Bea...   

                                                Plot  
0  A bartender is working at a saloon, serving dr...  
1  The moon, painted with a smiling face hangs ov...  
2  The film, just over a minute long, is composed...  
3  Lasting just 61 seconds and consisting of two ...  
4  The earliest known adaptation of the classic f...

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 34886 entries, 0 to 34885
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   Release Year      34886 non-null  int64
 1   Title             34886 non-null  str  
 2   Origin/Ethnicity  34886 non-null  str  
 3   Director          34886 non-null  str  
 4   Cast              33464 non-null  str  
 5   Genre             34886 non-null  str  
 6   Wiki Page         34886 non-null  str  
 7   Plot              34886 non-null  str  
dtypes: int64(1), str(7)
memory usage: 2.1 MB


In [5]:
df['Plot'][0]

"A bartender is working at a saloon, serving drinks to customers. After he fills a stereotypically Irish man's bucket with beer, Carrie Nation and her followers burst inside. They assault the Irish man, pulling his hat over his eyes and then dumping the beer over his head. The group then begin wrecking the bar, smashing the fixtures, mirrors, and breaking the cash register. The bartender then sprays seltzer water in Nation's face before a group of policemen appear and order everybody to leave.[1]"

In [6]:
df['Wiki Page'][0]

'https://en.wikipedia.org/wiki/Kansas_Saloon_Smashers'

In [7]:
df['Genre'].isna()

0        False
1        False
2        False
3        False
4        False
         ...  
34881    False
34882    False
34883    False
34884    False
34885    False
Name: Genre, Length: 34886, dtype: bool

In [8]:
df.isna().sum()

Release Year           0
Title                  0
Origin/Ethnicity       0
Director               0
Cast                1422
Genre                  0
Wiki Page              0
Plot                   0
dtype: int64

In [9]:
df.duplicated().sum()

np.int64(0)

In [10]:
df.describe()

Release Year
count  34886.000000
mean    1981.314252
std       27.815174
min     1901.000000
25%     1957.000000
50%     1988.000000
75%     2007.000000
max     2017.000000

In [11]:

df['Release Year'].value_counts()


Release Year
2013    1021
2014     929
2012     874
2011     858
2010     825
        ... 
1907       3
1903       2
1905       2
1902       1
1904       1
Name: count, Length: 117, dtype: int64

In [12]:

df['Title'].value_counts()


Title
Cinderella              8
The Three Musketeers    8
Treasure Island         7
Alice in Wonderland     6
Anna Karenina           6
                       ..
The Water Diviner       1
Çalgı Çengi İkimiz      1
Olanlar Oldu            1
Non-Transferable        1
İstanbul Kırmızısı      1
Name: count, Length: 32432, dtype: int64

In [13]:

df['Origin/Ethnicity'].value_counts()


Origin/Ethnicity
American        17377
British          3670
Bollywood        2931
Tamil            2599
Telugu           1311
Japanese         1188
Malayalam        1095
Hong Kong         791
Canadian          723
Australian        576
South_Korean      522
Chinese           463
Kannada           444
Bengali           306
Russian           232
Marathi           141
Filipino          128
Bangladeshi        87
Punjabi            84
Malaysian          70
Turkish            70
Egyptian           67
Assamese            9
Maldivian           2
Name: count, dtype: int64

In [14]:

df['Director'].value_counts()


Director
Unknown                    1124
Michael Curtiz               79
Hanna-Barbera                77
Lloyd Bacon                  66
Jules White                  63
                           ... 
Director: Kaan Müjdeci        1
Director: Russell Crowe       1
Selçuk Aydemir                1
Brendan Bradley               1
Ferzan Özpetek                1
Name: count, Length: 12593, dtype: int64

In [15]:

df['Cast'].value_counts()


Cast
Tom and Jerry                                                                                               80
Three Stooges                                                                                               56
The Three Stooges                                                                                           47
Looney Tunes                                                                                                37
Bugs Bunny                                                                                                  36
                                                                                                            ..
Director: Russell Crowe\r\nCast: Russell Crowe, Olga Kurylenko, Jai Courtney, Cem Yılmaz, Yılmaz Erdoğan     1
Ahmet Kural, Murat Cemcir                                                                                    1
Ata Demirer, Tuvana Türkay, Ülkü Duru                                                                      

In [16]:
df['Genre'].value_counts()

Genre
unknown                                      6083
drama                                        5964
comedy                                       4379
horror                                       1167
action                                       1098
                                             ... 
adventure, drama film                           1
sport film                                      1
animation, produced by glukoza production       1
adventure, romance, fantasy film                1
horror romantic comedy                          1
Name: count, Length: 2265, dtype: int64

In [17]:
df['Wiki Page'].value_counts()

Wiki Page
https://en.wikipedia.org/wiki/Himitsu_Sentai_Gorenger                        5
https://en.wikipedia.org/wiki/Digimon_Adventure_tri.                         5
https://en.wikipedia.org/wiki/Mani_Malai                                     4
https://en.wikipedia.org/wiki/Tamayura                                       4
https://en.wikipedia.org/wiki/Mob_Story                                      3
                                                                            ..
https://en.wikipedia.org/wiki/Sivas_(film)                                   1
https://en.wikipedia.org/wiki/%C3%87alg%C4%B1_%C3%87engi_%C4%B0kimiz         1
https://en.wikipedia.org/wiki/Olanlar_Oldu                                   1
https://en.wikipedia.org/wiki/Non-Transferable_(film)                        1
https://en.wikipedia.org/wiki/%C4%B0stanbul_K%C4%B1rm%C4%B1z%C4%B1s%C4%B1    1
Name: count, Length: 34070, dtype: int64

In [18]:
(df['Plot'].isna() == True).sum()


np.int64(0)

In [19]:
df1 = df.copy()


In [20]:
import numpy as np
df1.replace(['unknown','Unknown',''], np.nan, inplace=True)


Release Year                             Title Origin/Ethnicity  \
0              1901            Kansas Saloon Smashers         American   
1              1901     Love by the Light of the Moon         American   
2              1901           The Martyred Presidents         American   
3              1901  Terrible Teddy, the Grizzly King         American   
4              1902            Jack and the Beanstalk         American   
...             ...                               ...              ...   
34881          2014                 The Water Diviner          Turkish   
34882          2017                Çalgı Çengi İkimiz          Turkish   
34883          2017                      Olanlar Oldu          Turkish   
34884          2017                  Non-Transferable          Turkish   
34885          2017                İstanbul Kırmızısı          Turkish   

                                 Director  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4      George S. Fleming, Edwin S. Porter   
...                                   ...   
34881             Director: Russell Crowe   
34882                      Selçuk Aydemir   
34883                         Hakan Algül   
34884                     Brendan Bradley   
34885                      Ferzan Özpetek   

                                                    Cast            Genre  \
0                                                    NaN              NaN   
1                                                    NaN              NaN   
2                                                    NaN              NaN   
3                                                    NaN              NaN   
4                                                    NaN              NaN   
...                                                  ...              ...   
34881  Director: Russell Crowe\r\nCast: Russell Crowe...              NaN   
34882                          Ahmet Kural, Murat Cemcir           comedy   
34883              Ata Demirer, Tuvana Türkay, Ülkü Duru           comedy   
34884  YouTubers Shanna Malcolm, Shira Lazar, Sara Fl...  romantic comedy   
34885  Halit Ergenç, Tuba Büyüküstün, Mehmet Günsür, ...         romantic   

                                               Wiki Page  \
0      https://en.wikipedia.org/wiki/Kansas_Saloon_Sm...   
1      https://en.wikipedia.org/wiki/Love_by_the_Ligh...   
2      https://en.wikipedia.org/wiki/The_Martyred_Pre...   
3      https://en.wikipedia.org/wiki/Terrible_Teddy,_...   
4      https://en.wikipedia.org/wiki/Jack_and_the_Bea...   
...                                                  ...   
34881    https://en.wikipedia.org/wiki/The_Water_Diviner   
34882  https://en.wikipedia.org/wiki/%C3%87alg%C4%B1_...   
34883         https://en.wikipedia.org/wiki/Olanlar_Oldu   
34884  https://en.wikipedia.org/wiki/Non-Transferable...   
34885  https://en.wikipedia.org/wiki/%C4%B0stanbul_K%...   

                                                    Plot  
0      A bartender is working at a saloon, serving dr...  
1      The moon, painted with a smiling face hangs ov...  
2      The film, just over a minute long, is composed...  
3      Lasting just 61 seconds and consisting of two ...  
4      The earliest known adaptation of the classic f...  
...                                                  ...  
34881  The film begins in 1919, just after World War ...  
34882  Two musicians, Salih and Gürkan, described the...  
34883  Zafer, a sailor living with his mother Döndü i...  
34884  The film centres around a young woman named Am...  
34885  The writer Orhan Şahin returns to İstanbul aft...  

[34886 rows x 8 columns]

In [21]:
df1.isna().sum()

Release Year           0
Title                  2
Origin/Ethnicity       0
Director            1124
Cast                1423
Genre               6083
Wiki Page              0
Plot                   0
dtype: int64

In [22]:
df1.nunique()


Release Year          117
Title               32431
Origin/Ethnicity       24
Director            12592
Cast                32181
Genre                2264
Wiki Page           34070
Plot                33869
dtype: int64

In [23]:
title_missing = df[df1['Title'].isnull()]
print(title_missing)

       Release Year    Title Origin/Ethnicity                      Director  \
16136          2011  Unknown         American            Jaume Collet-Serra   
21110          2011  Unknown          British  Director: Jaume Collet-Serra   

                                                    Cast            Genre  \
16136  Liam Neeson, Diane Kruger, January Jones, Fran...  action thriller   
21110  Director: Jaume Collet-Serra\r\nCast: Liam Nee...          unknown   

                                               Wiki Page  \
16136  https://en.wikipedia.org/wiki/Unknown_(2011_film)   
21110  https://en.wikipedia.org/wiki/Unknown_(2011_film)   

                                                    Plot  
16136  Dr. Martin Harris and his wife Liz arrive in B...  
21110  Dr. Martin Harris and his wife Liz arrive in B...  


In [24]:

df1.dropna(subset=['Title'], inplace=True)

In [25]:
df1.reset_index(drop=True, inplace=True)

In [26]:
df1.isna().sum()

Release Year           0
Title                  0
Origin/Ethnicity       0
Director            1124
Cast                1423
Genre               6082
Wiki Page              0
Plot                   0
dtype: int64

### found that over 1124 entries in Director is null
### over 1423 entries in the cast is null
### over 6083 entries in the genre is null


In [27]:
import re

def clean_plot(text):
    text = text.lower()
    text = re.sub(r"\[\d+\]", "", text)  # remove [1], [2]
    text = re.sub(r"\{\{.*?\}\}", "", text)  # templates
    text = re.sub(r"==.*?==", "", text)  # headings
    text = re.sub(r"\s+", " ", text)
    return text.strip()


In [28]:
df1["clean_plot"] = df1["Plot"].apply(clean_plot)


In [29]:
df1.head()

Release Year                             Title Origin/Ethnicity  \
0          1901            Kansas Saloon Smashers         American   
1          1901     Love by the Light of the Moon         American   
2          1901           The Martyred Presidents         American   
3          1901  Terrible Teddy, the Grizzly King         American   
4          1902            Jack and the Beanstalk         American   

                             Director Cast Genre  \
0                                 NaN  NaN   NaN   
1                                 NaN  NaN   NaN   
2                                 NaN  NaN   NaN   
3                                 NaN  NaN   NaN   
4  George S. Fleming, Edwin S. Porter  NaN   NaN   

                                           Wiki Page  \
0  https://en.wikipedia.org/wiki/Kansas_Saloon_Sm...   
1  https://en.wikipedia.org/wiki/Love_by_the_Ligh...   
2  https://en.wikipedia.org/wiki/The_Martyred_Pre...   
3  https://en.wikipedia.org/wiki/Terrible_Teddy,_...   
4  https://en.wikipedia.org/wiki/Jack_and_the_Bea...   

                                                Plot  \
0  A bartender is working at a saloon, serving dr...   
1  The moon, painted with a smiling face hangs ov...   
2  The film, just over a minute long, is composed...   
3  Lasting just 61 seconds and consisting of two ...   
4  The earliest known adaptation of the classic f...   

                                          clean_plot  
0  a bartender is working at a saloon, serving dr...  
1  the moon, painted with a smiling face hangs ov...  
2  the film, just over a minute long, is composed...  
3  lasting just 61 seconds and consisting of two ...  
4  the earliest known adaptation of the classic f...

In [30]:
df1['clean_plot'][0]

"a bartender is working at a saloon, serving drinks to customers. after he fills a stereotypically irish man's bucket with beer, carrie nation and her followers burst inside. they assault the irish man, pulling his hat over his eyes and then dumping the beer over his head. the group then begin wrecking the bar, smashing the fixtures, mirrors, and breaking the cash register. the bartender then sprays seltzer water in nation's face before a group of policemen appear and order everybody to leave."

In [31]:
df1["Director"] = df1["Director"].fillna("").str.strip()
df1["Cast"] = df1["Cast"].fillna("").str.strip()
df1["Genre"] = df1["Genre"].fillna("").str.strip()
    

In [32]:
df1.isna().sum()

Release Year        0
Title               0
Origin/Ethnicity    0
Director            0
Cast                0
Genre               0
Wiki Page           0
Plot                0
clean_plot          0
dtype: int64

In [33]:


def normalize_cast(cast):
    if not isinstance(cast, str) or cast.strip() == "":
        return []
    return [c.strip() for c in cast.split(",") if c.strip()]




In [34]:
def normalize_genre(genre):
    if not isinstance(genre, str) or genre.strip() == "":
        return []
    return [g.strip().title() for g in genre.split(",") if g.strip()]



In [35]:
df1["cast_list"] = df1["Cast"].apply(normalize_cast)
df1["genre_list"] = df1["Genre"].apply(normalize_genre)


In [36]:
df1.head(10)

Release Year                             Title Origin/Ethnicity  \
0          1901            Kansas Saloon Smashers         American   
1          1901     Love by the Light of the Moon         American   
2          1901           The Martyred Presidents         American   
3          1901  Terrible Teddy, the Grizzly King         American   
4          1902            Jack and the Beanstalk         American   
5          1903               Alice in Wonderland         American   
6          1903           The Great Train Robbery         American   
7          1904                   The Suburbanite         American   
8          1905          The Little Train Robbery         American   
9          1905        The Night Before Christmas         American   

                             Director       Cast    Genre  \
0                                                           
1                                                           
2                                                           
3                                                           
4  George S. Fleming, Edwin S. Porter                       
5                      Cecil Hepworth  May Clark            
6                     Edwin S. Porter             western   
7                  Wallace McCutcheon              comedy   
8                Edwin Stanton Porter                       
9                Edwin Stanton Porter                       

                                           Wiki Page  \
0  https://en.wikipedia.org/wiki/Kansas_Saloon_Sm...   
1  https://en.wikipedia.org/wiki/Love_by_the_Ligh...   
2  https://en.wikipedia.org/wiki/The_Martyred_Pre...   
3  https://en.wikipedia.org/wiki/Terrible_Teddy,_...   
4  https://en.wikipedia.org/wiki/Jack_and_the_Bea...   
5  https://en.wikipedia.org/wiki/Alice_in_Wonderl...   
6  https://en.wikipedia.org/wiki/The_Great_Train_...   
7      https://en.wikipedia.org/wiki/The_Suburbanite   
8  https://en.wikipedia.org/wiki/The_Little_Train...   
9  https://en.wikipedia.org/wiki/The_Night_Before...   

                                                Plot  \
0  A bartender is working at a saloon, serving dr...   
1  The moon, painted with a smiling face hangs ov...   
2  The film, just over a minute long, is composed...   
3  Lasting just 61 seconds and consisting of two ...   
4  The earliest known adaptation of the classic f...   
5  Alice follows a large white rabbit down a "Rab...   
6  The film opens with two bandits breaking into ...   
7  The film is about a family who move to the sub...   
8  The opening scene shows the interior of the ro...   
9  Scenes are introduced using lines of the poem....   

                                          clean_plot    cast_list genre_list  
0  a bartender is working at a saloon, serving dr...           []         []  
1  the moon, painted with a smiling face hangs ov...           []         []  
2  the film, just over a minute long, is composed...           []         []  
3  lasting just 61 seconds and consisting of two ...           []         []  
4  the earliest known adaptation of the classic f...           []         []  
5  alice follows a large white rabbit down a "rab...  [May Clark]         []  
6  the film opens with two bandits breaking into ...           []  [Western]  
7  the film is about a family who move to the sub...           []   [Comedy]  
8  the opening scene shows the interior of the ro...           []         []  
9  scenes are introduced using lines of the poem....           []         []

In [37]:
df1[["Title", "Cast", "cast_list"]].sample(5)
df1[["Title", "Genre", "genre_list"]].sample(5)


Title     Genre  genre_list
11930             White Men Can't Jump    comedy    [Comedy]
10449               The Slugger's Wife   romance   [Romance]
3099              Ladies in Retirement                    []
28227  Chaappa Kurish (ചാപ്പാ കുരിശു്)  thriller  [Thriller]
4844                  Batman and Robin    serial    [Serial]

In [38]:
(df1["cast_list"].apply(len) == 0).sum()
(df1["genre_list"].apply(len) == 0).sum()


np.int64(6110)

In [39]:
df1.info()

<class 'pandas.DataFrame'>
RangeIndex: 34884 entries, 0 to 34883
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Release Year      34884 non-null  int64 
 1   Title             34884 non-null  str   
 2   Origin/Ethnicity  34884 non-null  str   
 3   Director          34884 non-null  str   
 4   Cast              34884 non-null  str   
 5   Genre             34884 non-null  str   
 6   Wiki Page         34884 non-null  str   
 7   Plot              34884 non-null  str   
 8   clean_plot        34884 non-null  str   
 9   cast_list         34884 non-null  object
 10  genre_list        34884 non-null  object
dtypes: int64(1), object(2), str(8)
memory usage: 2.9+ MB


In [40]:

df1["cast_list"].apply(type).value_counts()

cast_list
<class 'list'>    34884
Name: count, dtype: int64

In [41]:
df1["genre_list"].apply(type).value_counts()


genre_list
<class 'list'>    34884
Name: count, dtype: int64

In [42]:
df1.to_csv('df1_cleaned.csv', index=False)